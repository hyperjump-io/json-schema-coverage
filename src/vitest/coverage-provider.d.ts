import type { CoverageProviderModule } from "vitest";
import type { CustomProviderOptions } from "vitest/node";

export default CoverageProviderModule;

export type JsonSchemaCoverageProviderOptions = CustomProviderOptions & {
  provider: "custom";
  customProviderModule: "@hyperjump/json-schema-coverage/vitest-coverage-provider";
  include?: string[];
};
