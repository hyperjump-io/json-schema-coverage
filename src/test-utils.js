import { cwd } from "node:process";

/**
 * @import { CoverageMapData } from "istanbul-lib-coverage"
 */

/** @type (coverageMap: CoverageMapData) => CoverageMapData */
export const censorCoverageMap = (coverageMap) => {
  /** @type CoverageMapData */
  const censoredCoverageMap = {};
  for (const filePath in coverageMap) {
    const fileCoverageData = coverageMap[filePath];

    const censoredFilePath = filePath.slice(cwd().length);
    censoredCoverageMap[censoredFilePath] = fileCoverageData;

    for (const key in fileCoverageData) {
      switch (key) {
        case "path":
          fileCoverageData.path = censoredFilePath;
          break;
        case "statementMap":
        case "branchMap":
        case "fnMap":
        case "s":
        case "b":
        case "f":
          for (const schemaUri in fileCoverageData[key]) {
            if (schemaUri.startsWith("file:")) {
              const censoredSchemaUri = schemaUri.slice(cwd().length + 8);
              fileCoverageData[key][censoredSchemaUri] = fileCoverageData[key][schemaUri];
              delete fileCoverageData[key][schemaUri];

              if (key === "fnMap") {
                fileCoverageData[key][censoredSchemaUri].name = censoredSchemaUri;
              }
            }
          }
          break;
      }
    }
  }
  const [filePath] = Object.keys(coverageMap);
  const censoredFilePath = filePath.slice(cwd().length);
  coverageMap[filePath].path = censoredFilePath;
  return { [censoredFilePath]: coverageMap[filePath] };
};
