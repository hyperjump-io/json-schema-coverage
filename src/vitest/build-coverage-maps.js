import { existsSync } from "node:fs";
import { readFile } from "node:fs/promises";
import { join, resolve } from "node:path";
import { glob } from "tinyglobby";
import ignore from "ignore";
import { registerSchema } from "./index.js";
import { FileCoverageMapService } from "./file-coverage-map-service.js";

/**
 * @import { TestProject } from "vitest/node"
 */

/** @type (projects: TestProject) => Promise<void> */
export const setup = async (projects) => {
  const config = projects.vitest.config;

  const coverageService = new FileCoverageMapService(join(".json-schema-coverage", "maps"));

  // If --project filter is set pick only roots of resolved projects
  const roots = config.project?.length
    ? [...new Set(projects.vitest.projects.map((project) => project.config.root))]
    : [config.root];

  const include = "include" in config.coverage
    ? /** @type string[] */ (config.coverage.include)
    : ["**/*.schema.(json|yaml|yml)"];

  // Build coverage maps
  for (const root of roots) {
    const i = ignore();
    const gitignorePath = resolve(root, ".gitignore");
    if (existsSync(gitignorePath)) {
      const gitignore = await readFile(gitignorePath, "utf-8");
      i.add(gitignore);
    }

    let includedFiles = await glob(include, {
      cwd: root,
      dot: true,
      onlyFiles: true
    });

    const files = i
      .filter(includedFiles)
      .map((file) => resolve(root, file));

    for (const schemaPath of files) {
      try {
        await registerSchema(schemaPath);
      } catch (_error) {
      }
    }

    for (const file of files) {
      await coverageService.addFromFile(file);
    }
  }
};
