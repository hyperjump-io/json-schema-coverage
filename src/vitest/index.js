import "./json-schema-matcher.js";
