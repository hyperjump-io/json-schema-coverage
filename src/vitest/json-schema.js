import "@hyperjump/json-schema/draft-2020-12";
import "@hyperjump/json-schema/draft-2019-09";
import "@hyperjump/json-schema/draft-07";
import "@hyperjump/json-schema/draft-06";
import "@hyperjump/json-schema/draft-04";
import "@hyperjump/json-schema/openapi-3-0";
import "@hyperjump/json-schema/openapi-3-1";
import { buildSchemaDocument } from "@hyperjump/json-schema/experimental";
import { addMediaTypePlugin } from "@hyperjump/browser";
import contentTypeParser from "content-type";
import YAML from "yaml";

/**
 * @import { SchemaObject } from "@hyperjump/json-schema"
 */

addMediaTypePlugin("application/schema+json", {
  parse: async (response) => {
    const contentType = contentTypeParser.parse(response.headers.get("content-type") ?? "");
    const contextDialectId = contentType.parameters.schema ?? contentType.parameters.profile;

    /** @type SchemaObject */
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const json = await response.json();
    return buildSchemaDocument(json, response.url, contextDialectId);
  },
  // eslint-disable-next-line @typescript-eslint/require-await
  fileMatcher: async (path) => /\.json$/i.test(path)
});

addMediaTypePlugin("application/schema+yaml", {
  parse: async (response) => {
    const contentType = contentTypeParser.parse(response.headers.get("content-type") ?? "");
    const contextDialectId = contentType.parameters.schema ?? contentType.parameters.profile;

    /** @type SchemaObject */
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const yaml = YAML.parse(await response.text());
    return buildSchemaDocument(yaml, response.url, contextDialectId);
  },
  // eslint-disable-next-line @typescript-eslint/require-await
  fileMatcher: async (path) => /\.ya?ml$/i.test(path)
});

export * from "@hyperjump/json-schema/draft-2020-12";
