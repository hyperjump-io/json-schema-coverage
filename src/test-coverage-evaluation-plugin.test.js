import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, test } from "vitest";
import { mkdir, rm, writeFile } from "node:fs/promises";
import { resolve } from "node:path";
import { registerSchema, validate } from "@hyperjump/json-schema/draft-2020-12";
import { CoverageMapService, TestCoverageEvaluationPlugin } from "./index.js";
import { censorCoverageMap } from "./test-utils.js";

describe("TestCoverageEvaluationPlugin", () => {
  /** @type CoverageMapService */
  let coverageMapService;

  const fixtureDirectory = ".test-coverage-evaluation-plugin-fixtures";

  /** @type string */
  let fixturePath;

  /** @type (fixturePath: string, schemaJson: string) => Promise<void> */
  const addSchemaFile = async (fileName, schemaJson) => {
    fixturePath = resolve(fixtureDirectory, fileName);
    await writeFile(fixturePath, schemaJson);

    await coverageMapService.addFromFile(fixturePath);
  };

  beforeAll(async () => {
    await mkdir(fixtureDirectory, { recursive: true });
  });

  afterAll(async () => {
    await rm(fixtureDirectory, { recursive: true });
  });

  afterEach(async () => {
    await rm(fixturePath);
  });

  beforeEach(() => {
    coverageMapService = new CoverageMapService();
  });

  test("root schema", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("passing branching keyword", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "number"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("failing branching keyword", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "number"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, "foo", { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("both branches of a branching keyword", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "type": "number"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });
    await validate(fixturePath, "foo", { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("non-branching keyword", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "Foo"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("non-statement keyword", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$comment": "Foo"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("sub-schema", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "foo": true
  }
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, { foo: 42 }, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("sub-schema that isn't evaluated", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "foo": true
  }
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("embedded schema", async () => {
    await addSchemaFile("subject.schema.json", `{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "properties": {
    "foo": { "$ref": "#/$defs/foo" }
  },
  "$defs": {
    "foo": { "type": "number" }
  }
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, { foo: 42 }, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });

  test("reference to a non-file schema", async () => {
    registerSchema({ type: "number" }, "https://example.com/foo", "https://json-schema.org/draft/2020-12/schema");
    await addSchemaFile("subject.schema.json", `{
  "$id": "https://example.com/subject",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$ref": "/foo"
}`);
    const testCoveragePlugin = new TestCoverageEvaluationPlugin(coverageMapService);
    await validate(fixturePath, 42, { plugins: [testCoveragePlugin] });

    expect(censorCoverageMap(testCoveragePlugin.coverage)).toMatchSnapshot();
  });
});
