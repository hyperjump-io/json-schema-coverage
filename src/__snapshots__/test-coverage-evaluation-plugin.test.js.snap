// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TestCoverageEvaluationPlugin > both branches of a branching keyword 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": [
        1,
        1,
      ],
    },
    "branchMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "line": 3,
        "loc": {
          "end": {
            "column": 17,
            "line": 3,
          },
          "start": {
            "column": 2,
            "line": 3,
          },
        },
        "locations": [
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
        ],
        "type": "keyword",
      },
    },
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 2,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 4,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 2,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": 2,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 4,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "end": {
          "column": 17,
          "line": 3,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > embedded schema 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo/type": [
        0,
        1,
      ],
    },
    "branchMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo/type": {
        "line": 7,
        "loc": {
          "end": {
            "column": 28,
            "line": 7,
          },
          "start": {
            "column": 13,
            "line": 7,
          },
        },
        "locations": [
          {
            "end": {
              "column": 28,
              "line": 7,
            },
            "start": {
              "column": 13,
              "line": 7,
            },
          },
          {
            "end": {
              "column": 28,
              "line": 7,
            },
            "start": {
              "column": 13,
              "line": 7,
            },
          },
        ],
        "type": "keyword",
      },
    },
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 9,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo": {
        "decl": {
          "end": {
            "column": 8,
            "line": 7,
          },
          "start": {
            "column": 4,
            "line": 7,
          },
        },
        "line": 7,
        "loc": {
          "end": {
            "column": 30,
            "line": 7,
          },
          "start": {
            "column": 4,
            "line": 7,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo",
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "decl": {
          "end": {
            "column": 8,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "line": 4,
        "loc": {
          "end": {
            "column": 35,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo/type": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo/$ref": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 9,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo": {
        "end": {
          "column": 30,
          "line": 7,
        },
        "start": {
          "column": 4,
          "line": 7,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/$defs/foo/type": {
        "end": {
          "column": 28,
          "line": 7,
        },
        "start": {
          "column": 13,
          "line": 7,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": {
        "end": {
          "column": 2,
          "line": 5,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "end": {
          "column": 35,
          "line": 4,
        },
        "start": {
          "column": 4,
          "line": 4,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo/$ref": {
        "end": {
          "column": 33,
          "line": 4,
        },
        "start": {
          "column": 13,
          "line": 4,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > failing branching keyword 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": [
        1,
        0,
      ],
    },
    "branchMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "line": 3,
        "loc": {
          "end": {
            "column": 17,
            "line": 3,
          },
          "start": {
            "column": 2,
            "line": 3,
          },
        },
        "locations": [
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
        ],
        "type": "keyword",
      },
    },
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 4,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 4,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "end": {
          "column": 17,
          "line": 3,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > non-branching keyword 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 4,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/title": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 4,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/title": {
        "end": {
          "column": 15,
          "line": 3,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > non-statement keyword 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 4,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 4,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > passing branching keyword 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": [
        0,
        1,
      ],
    },
    "branchMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "line": 3,
        "loc": {
          "end": {
            "column": 17,
            "line": 3,
          },
          "start": {
            "column": 2,
            "line": 3,
          },
        },
        "locations": [
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
          {
            "end": {
              "column": 17,
              "line": 3,
            },
            "start": {
              "column": 2,
              "line": 3,
            },
          },
        ],
        "type": "keyword",
      },
    },
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 4,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 4,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/type": {
        "end": {
          "column": 17,
          "line": 3,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > reference to a non-file schema 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      "https://example.com/subject#": 1,
    },
    "fnMap": {
      "https://example.com/subject#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 5,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": "https://example.com/subject#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      "https://example.com/subject#": 1,
      "https://example.com/subject#/$ref": 1,
    },
    "statementMap": {
      "https://example.com/subject#": {
        "end": {
          "column": 0,
          "line": 5,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      "https://example.com/subject#/$ref": {
        "end": {
          "column": 15,
          "line": 4,
        },
        "start": {
          "column": 2,
          "line": 4,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > root schema 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 3,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 3,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > sub-schema 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 1,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 6,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "decl": {
          "end": {
            "column": 8,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "line": 4,
        "loc": {
          "end": {
            "column": 14,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 1,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 6,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": {
        "end": {
          "column": 2,
          "line": 5,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "end": {
          "column": 14,
          "line": 4,
        },
        "start": {
          "column": 4,
          "line": 4,
        },
      },
    },
  },
}
`;

exports[`TestCoverageEvaluationPlugin > sub-schema that isn't evaluated 1`] = `
{
  "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json": {
    "b": {},
    "branchMap": {},
    "f": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 0,
    },
    "fnMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "decl": {
          "end": {
            "column": 0,
            "line": 1,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "line": 1,
        "loc": {
          "end": {
            "column": 0,
            "line": 6,
          },
          "start": {
            "column": 0,
            "line": 1,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#",
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "decl": {
          "end": {
            "column": 8,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "line": 4,
        "loc": {
          "end": {
            "column": 14,
            "line": 4,
          },
          "start": {
            "column": 4,
            "line": 4,
          },
        },
        "name": ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo",
      },
    },
    "path": "/.test-coverage-evaluation-plugin-fixtures/subject.schema.json",
    "s": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": 1,
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": 0,
    },
    "statementMap": {
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#": {
        "end": {
          "column": 0,
          "line": 6,
        },
        "start": {
          "column": 0,
          "line": 1,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties": {
        "end": {
          "column": 2,
          "line": 5,
        },
        "start": {
          "column": 2,
          "line": 3,
        },
      },
      ".test-coverage-evaluation-plugin-fixtures/subject.schema.json#/properties/foo": {
        "end": {
          "column": 14,
          "line": 4,
        },
        "start": {
          "column": 4,
          "line": 4,
        },
      },
    },
  },
}
`;
